File: server/modules/arduino/arduino.py
import serial
import time

class Arduino:
    def __init__(self, serial_port, serial_speed, serial_timeout) -> None:
        self.serial_port = serial_port
        self.serial_speed = serial_speed
        self.serial_timeout = serial_timeout

        self.serial_connection = serial.Serial(
            port=self.serial_port,
            baudrate=self.serial_speed,
            timeout=self.serial_timeout
        )

        time.sleep(3)

    def send_message(self, message):
        self.serial_connection.write(message)

    def read_message(self):
        line = self.serial_connection.readline().decode('utf-8').strip()
        return line
    
    

File: src/modules/utils/arduino.js
import { SerialPort } from 'serialport';
import { ReadlineParser } from '@serialport/parser-readline';

export default class Arduino {
    constructor(serialPort, serialSpeed, serialTimeout) {
        this.serialPort = serialPort;
        this.serialSpeed = serialSpeed;
        this.serialTimeout = serialTimeout;

        this.port = new SerialPort({
            path: this.serialPort,
            baudRate: this.serialSpeed,
            autoOpen: true
        });

        this.parser = this.port.pipe(
            new ReadlineParser({ delimiter: '\n' })
        );

        this.ready = new Promise(resolve => setTimeout(resolve, 3000));
    }

    async sendMessage(message) {
        await this.ready;

        if (typeof message === 'string') {
            message = Buffer.from(message, 'utf-8');
        }

        return new Promise((resolve, reject) => {
            this.port.write(message, err => {
                if (err) return reject(err);
                resolve();
            });
        });
    }

    async readMessage() {
        await this.ready;

        return new Promise((resolve, reject) => {
            this.parser.once('data', data => {
                resolve(data.toString().trim());
            });
        });
    }
}